### PR TITLE
Git tokens

### PR DESCRIPTION
**Pull request by lukaszachy**
WDYT?

Needs #2223 to be merged first.
I've tried with my token and the plan:
```
execute:
    how: tmt
discover:
    how: fmf
    url: https://gitlab.com/redhat/rhel/tests/python
```

The NAME cannot be empty but its value is ignored (says gitlab docs)
```
$ TMT_GIT_TOKEN_URL_gl='gitlab.com/redhat/rhel' TMT_GIT_TOKEN_NAME_gl='foo' TMT_GIT_TOKEN_VALUE_gl='SECRET_TOKEN' tmt run discover
/var/tmp/tmt/run-065

/p
    discover
        how: fmf
        url: https://gitlab.com/redhat/rhel/tests/python
        summary: 1102 tests selected
```

Needs more docs and examples. Should be create teemtee private repo and token to test this or is unit test enough? 

* [x] implement the feature
* [x] write documentation
* [x] extend the test coverage
* [x] add a usage example
* [x] mention version

### CODE REVIEW

**Comment by lukaszachy** [comment]
I'll change to accept just two variables: url and token value. No need to complicate things with username when I can use token value as is (and document how it should look for gitlab/github).

The name might be confusing though, so ideas how to name it more than welcome. It will be "username+token  your forge expects in URL"

**Comment by happz** [comment]
> The name might be confusing though, so ideas how to name it more than welcome. It will be "username+token your forge expects in URL"

Credentials?

**Comment by lukaszachy** [comment]
> > The name might be confusing though, so ideas how to name it more than welcome. It will be "username+token your forge expects in URL"
> 
> Credentials?

/me facepalm.

**Comment by lukaszachy** [comment]
Ready for review.

**Comment by lukaszachy** [comment]
@psss How to solve 
```
       System Message: ERROR/3 (/tmp/tmppejve_cs/tmp/man.rst:, line 413)
              Unknown directive type "versionadded".
          .. versionadded:: 1.26
```

This directive works on readthedocs but rst2man seems to error on that.

**Comment by happz** [comment]
> @psss How to solve?
> 
> ```
>        System Message: ERROR/3 (/tmp/tmppejve_cs/tmp/man.rst:, line 413)
>               Unknown directive type "versionadded".
>           .. versionadded:: 1.26
> ```
> 
> This directive works on readthedocs but rst2man seems to error on that.
> 
> For now ([b788464](https://github.com/teemtee/tmt/commit/b7884646c30e6a7ef91d2d3ad42a91f0980c296b)) I'll remove that line from man.rst

I have a couple of changes for `README.rst` waiting for rhel8 to be gone. This requires a python package in buildtime for proper rendering of `versionadded` (or `code-block` in my case), but last time I checked, epel8 had a version way too old. So I'm waiting...

Maybe you could comment it out with a TODO note for later, `.. TODO: ..versionadded:: 1.26`.

**Comment by lukaszachy** [comment]
> Maybe you could comment it out with a TODO note for later, `.. TODO: ..versionadded:: 1.26`.

Hm, I'd prefer to have versionadded on readthedocs so I'll keep the 'sed' in Makefile .

**Comment by happz** [comment]
> > Maybe you could comment it out with a TODO note for later, `.. TODO: ..versionadded:: 1.26`.
> 
> Hm, I'd prefer to have versionadded on readthedocs so I'll keep the 'sed' in Makefile .

Ok. Add the `TODO` label there though, it's how I look for things to fix when I'm bored :)

**Comment by lukaszachy** [comment]
Rebased on current main and squashed. Kindly asking for review.

File: README.rst [hunk 1]
```diff
@@ -389,6 +389,29 @@ TMT_SHOW_TRACEBACK
     traceback is not printed out. When ``TMT_SHOW_TRACEBACK`` is
     set to any string except ``0``, traceback would be printed out.
 
+TMT_GIT_CREDENTIALS_URL\_, TMT_GIT_CREDENTIALS_VALUE\_
```
Comment (thrix) [inline README.rst:392]: Would it not be here better to include the suffix, as it is in the text?

```suggestion
TMT_GIT_CREDENTIALS_URL_<suffix>, TMT_GIT_CREDENTIALS_VALUE_<suffix>
```

File: README.rst [hunk 2]
```diff
@@ -389,6 +389,29 @@ TMT_SHOW_TRACEBACK
...
+    Prefixes of variable pairs used to provide credentials to clone git
```
Comment (thrix) [inline README.rst:393]: then we can maybe omit `Prefixes` from here?

File: tmt/utils.py [hunk 3]
```diff
@@ -2901,6 +2901,38 @@ def rewrite_git_url(url: str, patterns: List[Tuple[str, str]]) -> str:
...
+            # Get credentials value
+            try:
+                creds = os.environ[f'{INJECT_CREDENTIALS_VALUE_PREFIX}{unique_suffix}']
+            except KeyError:
+                raise GitUrlError(f'Missing variable with credentials for "{url}"')
+            # Return original url if credentials is an empty value
+            if not creds:
+                return url
+            # Finally inject credentials into the url and return it
+            return re.sub(r'([^/]+://)([^/]+)', r'\1' + f'{creds}@' + r'\2', url)
```
Comment (thrix) [inline tmt/utils.py:2931]: Cannot the replacement string be written together? Is the concatenation needed?

```suggestion
            return re.sub(r'([^/]+://)([^/]+)', rf'\1{creds}@\2', url)
```

**Review by thrix — CHANGES_REQUESTED** [review]

File: README.rst (hunk 1 above)
Comment (lukaszachy) [reply README.rst]: fixed in e7ac97ca

File: README.rst (hunk 2 above)
Comment (lukaszachy) [reply README.rst]: fixed in e7ac97ca

File: tmt/utils.py (hunk 3 above)
Comment (lukaszachy) [reply tmt/utils.py]: fixed in e7ac97ca

File: tests/unit/test_utils.py [hunk 4]
```diff
@@ -144,6 +145,40 @@ def test_clonable_git_url():
         == 'git://example.com'
 
 
+def test_inject_auth_git_url(monkeypatch) -> None:
```
Comment (thrix) [inline tests/unit/test_utils.py:148]: @lukaszachy ah missed this on first  review, could we add a test for this exception please?

```
                raise GitUrlError(f'Missing variable with credentials for "{url}"')
```

**Review by happz — APPROVED** [review]

File: tests/unit/test_utils.py (hunk 4 above)
Comment (lukaszachy) [reply tests/unit/test_utils.py:148]: fixed in 93a1973 and improved in d7b903f

Now it should state which variable is missing, eg.
`Missing "TMT_GIT_CREDENTIALS_VALUE_gl" variable with credentials for "https://gitlab.com/....."

**Review by thrix — APPROVED** [review]
LGTM

**Pull request merged**